### PR TITLE
Added support for importing csv data uris [#138395609]

### PR DESCRIPTION
**Pull request by dougmartin**

### CODE REVIEW

File: apps/dg/english.lproj/strings.js [hunk 1]
```diff
@@ -629,6 +629,7 @@ SC.stringsFor('English', {
   'DG.Inspector.newAttribute': "New Attribute in %@...",  // "New Attribute in %@..."
   'DG.Inspector.randomizeAllAttributes': "Rerandomize All", // "Randomize Attributes"
   'DG.Inspector.exportCaseData': "Export Case Data...", // "Export Case Data..."
+  'DG.Inspector.exportPrimaryCaseData.toolTip': "Export data from this case", // "Export Case Data..."
```
Comment (kswenson) [inline apps/dg/english.lproj/strings.js]: It's exporting more than just this case, so perhaps something like "Export data from this case table" or "Export cases as .csv".

**Comment by dougmartin** [comment]
@jsandoe @kswenson Should I split this PR so that the csv import gets merged?  Evangeline is testing the "sharinator" now which needs that functionality.

File: apps/dg/english.lproj/strings.js (hunk 1 above)
Comment (kswenson) [inline apps/dg/english.lproj/strings.js]: After review I figured Jonathan would handle merging it. I apparently didn't assign it to him in Github, however. I'll check with him to see how he wants to handle it.

**Comment by dougmartin** [comment]
@jsandoe @kswenson I've removed the export icon code from this PR

**Comment by jywarren** [comment]
Hi, this seems pretty interesting! Is there a way to initiate a new document (forgive my not knowing the nomenclature) or instance of Codap with an existing CSV file using this method? We are hoping to make it possible to initiate a Codap session using a URL something like:

`https://codap.concord.org/import?csv=https://example.com/data.csv`

Or, using this method perhaps, using a CSV data-url from another application. Is there a functionality like this?

**Comment by jsandoe** [comment]
Hello Jeffrey,

In fact there is a capability similar to what you are suggesting, though it wasn't explicitly provided by this PR, and when I tried it just now, I find it is broken. The call would be very similar to your suggestion: 

```https://codap.concord.org/import?url=https://example.com/data.csv``` 

I think a recent check-in broke the capability. I will log a bug, and post here when it is fixed.

**Comment by jsandoe** [comment]
Hello Jeffrey,

Here is a URL so you can see how it worked before the bug appeared:
https://codap.concord.org/releases/build_0481/static/dg/en/cert/index.html?url=https://data.cityofnewyork.us/api/views/kku6-nxdu/rows.csv?accessType=DOWNLOAD

**Comment by jywarren** [comment]
Awesome! @IshaGupta18 check it out!

**Comment by IshaGupta18** [comment]
Oh this looks good! Thanks a lot @jywarren @jsandoe !

**Comment by IshaGupta18** [comment]
@jsandoe I am sorry to bother you, but could you share a link to how `https://codap.concord.org/releases/build_0481/static/dg/en/cert/index.html?url=https://data.cityofnewyork.us/api/views/kku6-nxdu/rows.csv?accessType=DOWNLOAD` can be implemented in the docs? Thanks a ton.

**Comment by IshaGupta18** [comment]
Also, I think `https://github.com/concord-consortium/codap-data-interactives/tree/master/LiveDataSet` is similar to what we want to implement. Can anyone guide me to the docs through which we can produce a URL like above ie open an uploaded CSV in CODAP? @jywarren check this one out! Thanks a lot!

**Comment by IshaGupta18** [comment]
@jsandoe I get this error when I replace the URL parameter of `https://codap.concord.org/releases/build_0481/static/dg/en/cert/index.html?url=https://data.cityofnewyork.us/api/views/kku6-nxdu/rows.csv?accessType=DOWNLOAD`
with another remote file's link like this one: `http://insight.dev.schoolwires.com/HelpAssets/C2Assets/C2Files/C2ImportCalEventSample.csv`.
Is this a bug, or is there some other procedure to achieve similar results? If it's a bug, could you help me out with the solution? Thanks a ton!
image

**Comment by jsandoe** [comment]
Hello Isha,

The issue is that the link is to an insecure http address and CODAP is running with an https URL. Browser rules (quite reasonably) prevent a secure page from loading an insecure file. I tried modifying the URL to "https://insight.dev.schoolwires.com/HelpAssets/C2Assets/C2Files/C2ImportCalEventSample.csv" but the server doesn't appear to support https.

If you are affiliated with the owners of schoolwires.com in any way you could request that they support HTTPS. It's probably a good idea for them anyway.

The other network bugaboo that limits CODAP's ability to read data from a third party sources are same origin policies (CORS). Servers can be configured to restrict the ability of browser apps to read their data. If a CSV file's server is configured in this way, CODAP will not be able to use it directly.

The fallback, if you are unable to resolve either of these situations, is to down load the file to your laptop, then drag the file into a CODAP instance. You can see that demonstrated here: https://codap.concord.org/releases/latest/static/dg/en/cert/index.html#file=examples:Getting%20started%20with%20CODAP .

Thanks,

Jonathan

**Comment by jsandoe** [comment]
Hello @IshaGupta18 ,
I just noticed there were a couple of inquiries from you from 19 days ago that I hadn't seen before. I am not sure I understand what "implemented in the docs" means in this context. Does it mean, how would one create a CODAP document that would have the same state as the result of invoking "https://codap.concord.org/releases/build_0481/static/dg/en/cert/index.html?url=https://data.cityofnewyork.us/api/views/kku6-nxdu/rows.csv?accessType=DOWNLOAD"? If so that is simple enough. Simple click on the CODAP file menu in the upper left corner and select "Save..." or "Share...". "Save..." will allow you to create a document on Google drive or on your local file system for your later reference. "Share..." will create a URL for a version of the current state that you can easily share with others.
If you have further questions, perhaps it would be more efficient to set up a brief video meeting. I am sure we could resolve your remaining questions about CODAP pretty quickly.

Jonathan

**Comment by IshaGupta18** [comment]
Yes, this is what I aim to do, but through your API. Basically, how to create this link `https://codap.concord.org/releases/build_0481/static/dg/en/cert/index.html?url=https://data.cityofnewyork.us/api/views/kku6-nxdu/rows.csv?accessType=DOWNLOAD` through an API call. By `docs` I meant, where exactly can I find this
> Simple click on the CODAP file menu in the upper left corner and select "Save..." or "Share...". "Save..." will allow you to create a document on Google drive or on your local file system for your later reference. "Share..." will create a URL for a version of the current state that you can easily share with others.

in the docs to be done through an API call.

We most definitely can have a brief video call meeting to quickly clear up all the doubts and begin the implementation. I would really appreciate it. Let's schedule one soon.

Thanks a ton for your help! 
Isha Gupta

**Comment by jsandoe** [comment]
Ah, I see. We currently don't support document operations of that type through the Plugin API. There may be other ways to accommodate the need, though. My email address is jsandoe@concord.org. Please contact me there and we can set up a Zoom meeting.

Thanks,
Jonathan

**Comment by IshaGupta18** [comment]
Oh lovely, I would contact you there. Also, we can probably fix up a time at which @jywarren may also be able to join us, if it's fine with him?

**Pull request merged**